core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,301 @@
+ * "spooling" fields as they arrive by queueing the configured {@linkplain BiConsumer}, parsed value pair. Once all constructor arguments
+ * arrive it builds the target object and "unspools" the fields by draining the queue. Any fields that come in after the last constructor
+ * arguments are parsed just like {@linkplain ObjectParser}.
+ * </p>
+ * <p>
+ * Here is how you declare an ObjectSpooler:
+ * </p>
+ * <pre>{@code
+ *   private static final ObjectSpooler<Thing, Void> PARSER = new ObjectSpooler<>("thing",
+ *           a -> new Thing((String) a[0], (String) a[1]));
+ *   static {
+ *       PARSER.declareString(constructorArg(), new ParseField("animal"));
+ *       PARSER.declareString(constructorArg(), new ParseField("vegetable"));
+ *       PARSER.declareInt(Thing::setMineral, new ParseField("mineral"));
+ *       PARSER.declareInt(Thing::setFruit, new ParseField("fruit"));
+ *   }
+ * }</pre>
+ * <p>
+ * This does add some overhead compared to just using {@linkplain ObjectParser} directly. On a 2.2 GHz Intel Core i7 MacBook Air running on
+ * battery power in a reasonably unscientific microbenchmark it is about 100 microseconds for a reasonably large object, less if the
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + (2 * param_count)</code> objects per parse, plus a few ArrayList resizes for very long lists. If this overhead
+ * is too much for you then feel free to have ObjectParser parse a secondary object and have that one call the target object's constructor.
+ * </p>
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
+    /**
+     * Consumer that marks a field as a constructor argument instead of a real object field.
+     */
+    private static final BiConsumer<Object, Object> CONSTRUCTOR_ARG_MARKER = (a, b) -> {
+        throw new UnsupportedOperationException("I am just a marker I should never be called.");
+    };
+
+    /**
+     * Maximum initial size of the list of spooled fields. The actual initial size is bounded by the number of fields. If there happen to be
+     * more than this many fields then the initial size clamps to this value. The idea is that in most cases it is rare for a user to have
+     * to spool lots of fields. They'll either not send that many fields or send the required fields before they get to this many fields.
+     * The actual value is fairly arbitrary but feels like "less than lots but plenty more spooling than we expect on average."
+     */
+    private static final int MAX_INITIAL_SIZE_OF_SPOOLED_FIELDS = 6;
```

Comment:
I think this is an unneeded optimization, and I'd rather just avoid the resize reallocations?